File: app/src/main/java/com/genonbeta/TrebleShot/task/FindWorkingNetworkTask.java
/*
 * Copyright (C) 2020 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package com.genonbeta.TrebleShot.task;

import android.content.Context;
import android.util.Log;
import androidx.annotation.Nullable;
import com.genonbeta.TrebleShot.R;
import com.genonbeta.TrebleShot.object.Device;
import com.genonbeta.TrebleShot.object.DeviceAddress;
import com.genonbeta.TrebleShot.protocol.communication.CommunicationException;
import com.genonbeta.TrebleShot.service.backgroundservice.AttachableAsyncTask;
import com.genonbeta.TrebleShot.service.backgroundservice.AttachedTaskListener;
import com.genonbeta.TrebleShot.service.backgroundservice.TaskStoppedException;
import com.genonbeta.TrebleShot.util.AppUtils;
import com.genonbeta.TrebleShot.util.CommonErrorHelper;
import com.genonbeta.TrebleShot.util.CommunicationBridge;
import com.genonbeta.TrebleShot.util.Transfers;
import com.genonbeta.android.database.exception.ReconstructionFailedException;

import java.io.IOException;
import java.net.SocketException;
import java.util.List;

public class FindWorkingNetworkTask extends AttachableAsyncTask<FindWorkingNetworkTask.CalculationResultListener>
{
    private final Device device;

    public FindWorkingNetworkTask(Device device)
    {
        this.device = device;
    }

    @Override
    protected void onRun() throws TaskStoppedException
    {
        try {
            List<DeviceAddress> knownAddressList = AppUtils.getKuick(getContext()).castQuery(
                    Transfers.createAddressSelection(device.uid), DeviceAddress.class);

            progress().addToTotal(knownAddressList.size());
            publishStatus();

            if (knownAddressList.size() > 0) {
                for (DeviceAddress address : knownAddressList) {
                    throwIfStopped();
                    setOngoingContent(address.getHostAddress());
                    progress().addToCurrent(1);
                    publishStatus();

                    try (CommunicationBridge client = CommunicationBridge.connect(kuick(), address, device, 0)) {
                        client.requestAcquaintance();
                        if (client.receiveResult()) {
                            CalculationResultListener anchor = getAnchor();
                            if (anchor != null)
                                post(() -> anchor.onCalculationResult(device, address));
                            return;
                        }
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            }

            CalculationResultListener anchor = getAnchor();
            if (anchor != null)
                post(() -> anchor.onCalculationResult(device, null));
        } catch(Exception e) {
            post(CommonErrorHelper.messageOf(getContext(), e));
        }
    }

    @Override
    public String getName(Context context)
    {
        return context.getString(R.string.text_findAvailableNetwork);
    }

    public interface CalculationResultListener extends AttachedTaskListener
    {
        void onCalculationResult(Device device, @Nullable DeviceAddress address);
    }
}

File: app/src/main/java/com/genonbeta/TrebleShot/task/FindWorkingNetworkTask.kt
/*
 * Copyright (C) 2020 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.TrebleShot.task

import android.content.*
import com.genonbeta.TrebleShot.R
import com.genonbeta.TrebleShot.model.Device
import com.genonbeta.TrebleShot.model.DeviceAddress
import com.genonbeta.TrebleShot.service.backgroundservice.AttachableAsyncTask
import com.genonbeta.TrebleShot.service.backgroundservice.AttachedTaskListener
import com.genonbeta.TrebleShot.service.backgroundservice.TaskStoppedException
import com.genonbeta.TrebleShot.task.FindWorkingNetworkTask.CalculationResultListener
import com.genonbeta.TrebleShot.util.*
import com.genonbeta.TrebleShot.util.CommunicationBridge.Companion.receiveResult
import java.io.IOException

class FindWorkingNetworkTask(private val device: Device) : AttachableAsyncTask<CalculationResultListener>() {
    @Throws(TaskStoppedException::class)
    override fun onRun() {
        try {
            val knownAddressList: List<DeviceAddress> = AppUtils.getKuick(context).castQuery(
                Transfers.createAddressSelection(device.uid), DeviceAddress::class.java
            )
            progress.increaseTotalBy(knownAddressList.size)
            if (knownAddressList.isNotEmpty()) {
                for (address in knownAddressList) {
                    throwIfStopped()
                    ongoingContent = address.hostAddress
                    progress.increaseBy(1)
                    try {
                        CommunicationBridge.connect(kuick, address, device, 0).use { client ->
                            client.requestAcquaintance()
                            if (client.receiveResult()) {
                                post { anchor?.onCalculationResult(device, address) }
                                return
                            }
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }
            post { anchor?.onCalculationResult(device, null) }
        } catch (e: Exception) {
            e.printStackTrace()
            post(CommonErrorHelper.messageOf(context, e))
        }
    }

    override fun getName(context: Context): String {
        return context.getString(R.string.text_findAvailableNetwork)
    }

    interface CalculationResultListener : AttachedTaskListener {
        fun onCalculationResult(device: Device, address: DeviceAddress?)
    }
}